---
title: "rRNA pseudoU"
author: "Chris Katanski"
date: "4/16/2021"
output: html_document
---

#Set up
```{r, set up packages etc., echo=FALSE, results="hide", warning=FALSE, message=FALSE}

#Improt packages for plotting and data manipulation.
library(ggplot2)
library(scales)
library(reshape2)
library(tidyr)
library(dplyr, warn.conflicts = FALSE) #its important that this is after tidyr I think
# Suppress summarise info
options(dplyr.summarise.inform = FALSE)

library(grid)
library(gridExtra)
library(ggpubr)
library(ggrepel)
library(ggh4x) #For nested faceting

library(mixtools)

#Reading in svg files into r for ggplotting
#library(grImport2)
#library(grConvert)
#Reading and plotting jpeg and PNG
#library(jpeg)
#library(png)

#library(ggrepel)
library(stringr)
#library(forcats)
#library(readxl)

#For nice log labels
#library(cat.extras)
#library(plotly)

#library(extrafont)
#font_import()
#loadfonts()

#Set the global figure theme now
theme_set(
  theme_bw() + theme(legend.position = "bottom", 
        text = element_text(family = "Arial", size = 14),
        panel.grid.major = element_blank(),
        panel.grid.minor = element_blank(),
        panel.background = element_rect(color = "black"))
  )

#Define a function that excludes things
`%nin%` = Negate(`%in%`)

empty_graph <- ggplot() + theme_void()
```

#Set working driectory
```{r}
#CHANGE THIS TO MATCH YOUR COMPUTER
setwd("~/4SR/data/20210416_pseudoU_rRNA/source/")

```


#_
#Read in basewise
```{r}
#Noah, you might have to change this folder name
PATH = "../5_tsv/Assorted_rRNA/"
FILES <-data.frame( file_name = list.files(PATH) )

#Split generic file names into salient parts
FILES <- FILES %>%
  extract(file_name, "(.*).tsv", into="file_name2", remove=FALSE) %>% #"extract" is used for 'regular expressions'
  separate(file_name2, remove=TRUE, sep="_", c("library","junk1","junk2","barcode", "junk3","bin_start","bin_stop","junk4") ) %>% #"Separate" splits one column into several. Here I'm splitting the file name at underscore characters and putting the data into several columns like library and barcode
  mutate(sample = barcode) %>% #"mutate" is for making new columns or overwirting existing columns
  select(-barcode) %>% #"select" lets you pick which columns to keep. If you use a minus sign, then its 'keep everything, but drop this one'
  select(-junk1, -junk3, -junk4, -junk2) 

#PROJECT SPECIFIC DECODING
FILES$file_name <- as.character(FILES$file_name) #"Character" type is like a string in python. The contrast is "factor" for categorical data (google it)
FILES$bin_start <- as.character(FILES$bin_start)
FILES$bin_stop <- as.character(FILES$bin_stop)
FILES$library <- recode(FILES$library, 
                        "TP-CK-14S-WZ1"="plusCMC",
                        "TP-CK-14S-WZ2"="minusCMC",
                        "TP-CK-14S-WZ3"="bisulfite") #Recode lets you easily change a bunch of values all at once

#Chance the "library" column to DM
FILES <- FILES %>%
  mutate(treatment = library) %>%
  select(-library)

#=======================+
#Some fun fragment stuff. 
#The minus 3 bin is ALL READS. By contrast the minus 2 bin is all sense reads, whiles minus 1 is all antisense reads. 
FILES <- FILES %>%
  mutate(bin_start = ifelse(is.na(bin_start), -3, bin_start),
         bin_stop = ifelse(is.na(bin_stop), -3, bin_stop))
#========================

#A function that reads in a data file and adds above annotation
read_in_one <- function(row){
  output <- read.csv(paste0(PATH, row$file_name), header=T, sep="\t") %>%
    mutate(bin_start = row$bin_start,
           bin_stop  = row$bin_stop,
           sample = row$sample,
           treatment = row$treatment)
  #specify data types since enpty dataframes get confused and sad
  output$gene <- as.character(output$gene)
  output$base <- as.character(output$base)
  output$treatment <- as.character(output$treatment)
  return(output)
}


#Up to this point the "FILES" data frame is just identifying info, no actual tRNA data.
#The last step here is to run the funciton above such that it reads in the data in file with 
#"file_name" and identifies all of that data with the right patient, tissue, DM, bin_start, etc data

#A loop to read in all the data files and add the annotation
data_rRNA_assorted <- FILES %>%
  group_by(file_name) %>%
  do(read_in_one(.)) %>%
  ungroup()

```

#Wrtie excel for Tao
```{r}

write_data_to_excel <- function(row){
  PATH = paste0("/Users/ckatanski/4SR/data/20210416_pseudoU_rRNA/source/", row$treatment,"_", row$gene, ".csv")
  print(PATH)
  write.table(row$data, file=PATH, sep=",", row.names = F)
  return(data.frame(A=c(PATH)))
}


temp <- data_rRNA_assorted %>%
  ungroup() %>%
  select(-bin_start, -bin_stop, -file_name, -sample) %>%
  nest(data = c(-gene, -treatment)) %>%
  group_by(gene, treatment) %>%
  do(., write_data_to_excel(.))

```


#Write excel for Sihao
```{r}
temp <- data_rRNA_assorted %>%
  mutate(deletion_rate = deletion / pileup) %>%
  filter(treatment %in% c("minusCMC", "bisulfite")) %>%
  select(gene, position, base, treatment, pileup, deletion_rate) %>%
  pivot_wider(names_from = c("treatment"), values_from = c("pileup","deletion_rate"))

temp2 <- temp %>%
  filter(deletion_rate_bisulfite - deletion_rate_minusCMC >= 0.2) %>%
  filter(base=="T") %>%
  filter(pileup_minusCMC >=500, pileup_bisulfite >= 500)

write.table(temp2, col.names = T, sep=",", row.names = F,
            file="./model_organism_pseudoU.csv")

```

#_
#Read in basewise microbes
```{r}
#Noah, you might have to change this folder name
PATH = "../5_tsv/rfam_bacterial_rRNA_combine/"
FILES <-data.frame( file_name = list.files(PATH) )

#Split generic file names into salient parts
FILES <- FILES %>%
  extract(file_name, "(.*).tsv", into="file_name2", remove=FALSE) %>% #"extract" is used for 'regular expressions'
  separate(file_name2, remove=TRUE, sep="_", c("library","junk1","junk2","barcode", "junk3","bin_start","bin_stop","junk4") ) %>% #"Separate" splits one column into several. Here I'm splitting the file name at underscore characters and putting the data into several columns like library and barcode
  mutate(sample = barcode) %>% #"mutate" is for making new columns or overwirting existing columns
  select(-barcode) %>% #"select" lets you pick which columns to keep. If you use a minus sign, then its 'keep everything, but drop this one'
  select(-junk1, -junk3, -junk4, -junk2) 

#PROJECT SPECIFIC DECODING
FILES$file_name <- as.character(FILES$file_name) #"Character" type is like a string in python. The contrast is "factor" for categorical data (google it)
FILES$bin_start <- as.character(FILES$bin_start)
FILES$bin_stop <- as.character(FILES$bin_stop)
FILES$library <- recode(FILES$library, 
                        "TP-CK-14S-WZ1"="plusCMC",
                        "TP-CK-14S-WZ2"="minusCMC",
                        "TP-CK-14S-WZ3"="bisulfite") #Recode lets you easily change a bunch of values all at once

#Chance the "library" column to DM
FILES <- FILES %>%
  mutate(treatment = library) %>%
  select(-library)

#=======================+
#Some fun fragment stuff. 
#The minus 3 bin is ALL READS. By contrast the minus 2 bin is all sense reads, whiles minus 1 is all antisense reads. 
FILES <- FILES %>%
  mutate(bin_start = ifelse(is.na(bin_start), -3, bin_start),
         bin_stop = ifelse(is.na(bin_stop), -3, bin_stop))
#========================

#A function that reads in a data file and adds above annotation
read_in_one <- function(row){
  output <- read.csv(paste0(PATH, row$file_name), header=T, sep="\t") %>%
    mutate(bin_start = row$bin_start,
           bin_stop  = row$bin_stop,
           sample = row$sample,
           treatment = row$treatment)
  #specify data types since enpty dataframes get confused and sad
  output$gene <- as.character(output$gene)
  output$base <- as.character(output$base)
  output$treatment <- as.character(output$treatment)
  return(output)
}


#Up to this point the "FILES" data frame is just identifying info, no actual tRNA data.
#The last step here is to run the funciton above such that it reads in the data in file with 
#"file_name" and identifies all of that data with the right patient, tissue, DM, bin_start, etc data

#A loop to read in all the data files and add the annotation
data_rRNA_microbe <- FILES %>%
  group_by(file_name) %>%
  do(read_in_one(.)) %>%
  ungroup()

```

#Extra processing for names and subunit
```{r}
rRNA_microbe_data <- data_rRNA_microbe %>%
  separate(gene, sep="SU_", remove=F, into=c("subunit","junk")) %>%
  select(-junk) %>%
  select(-file_name, -bin_start, -bin_stop, -sample)

```

#Get the interesting sites
```{r}
temp <- rRNA_microbe_data %>%
  mutate(deletion_rate = deletion / pileup) %>%
  filter(treatment %in% c("minusCMC", "bisulfite")) %>%
  select(gene, subunit, position, base, treatment, pileup, deletion_rate) %>%
  pivot_wider(names_from = c("treatment"), values_from = c("pileup","deletion_rate"))

temp2 <- temp %>%
  filter(deletion_rate_bisulfite - deletion_rate_minusCMC >= 0.2) %>%
  filter(base=="T") %>%
  filter(pileup_minusCMC >=50, pileup_bisulfite >= 50)

write.table(temp2, col.names = T, sep=",", row.names = F,
            file="./rfam_bacteria_pseudoU.csv")

```


#_
#match up names to taxonomy
##Read in NCBI taxon info
```{r}

PATH="../ncbi_taxonomy/Organismal_taxonomy/taxdump 2/nodes.dmp"
Nodes = read.csv(PATH, sep="\t", header=F) 
Nodes <- Nodes %>%
  select(V1, V3, V5)
colnames(Nodes) <- c("taxonID", "parentID","Rank")
Nodes <- Nodes %>%
  filter(Rank=="no rank") %>%
  mutate(Rank="no_rank") %>%
  rbind(filter(Nodes, Rank!="no rank")) %>%
  mutate(taxonID = as.numeric(taxonID),
         parentID = as.numeric(parentID),
         Rank = as.factor(Rank))

PATH="../ncbi_taxonomy/Organismal_taxonomy/taxdump 2/names.dmp"
Names = read.csv(PATH, sep="\t", header=F) 
Names <- Names %>%
  select(V1, V3, V7) %>%
  filter(V7 =="scientific name") %>%
  select(-V7) %>%
  mutate(V1 = as.numeric(V1))
colnames(Names) <- c("taxonID","name")
```

###Make big taxon table
```{r}

recursive_taxon3 <- function(df, Nodes, Names, rank="species"){
  link = filter(Nodes, taxonID %in% df$taxonID)
  temp_data <- df %>%
    full_join(., link, by=c("taxonID"))
  #Split into matches and mismatches
  matches = filter(temp_data, Rank==rank)
  mismatches = filter(temp_data, Rank != rank) %>%
    mutate(taxonID=parentID) %>%
    select(-parentID, -Rank)

  #Check for root; update mismatches
  if(identical(unique(link$taxonID), c(1) )  ){
    print(paste0("root"," ", rank))
    mismatches[rank] = "skip"
    mismatches[paste0(rank, "_ID")] = 0
    mismatches$taxonID = mismatches$OG_ID
    }

  #For non-matches, update taxonID, recursive
  else{ if(nrow(df >0 )){
    mismatches <- recursive_taxon3(mismatches, Nodes, Names, rank) } }
  
  #If matches is NOT empty, update it
  if(nrow(matches != 0 )){
    #Update matches
    return_df = matches %>%
      full_join(., Names, by=c("taxonID")) %>%
      filter(taxonID %in% matches$taxonID)
    
    return_df[rank] = return_df["name"]
    return_df[paste0(rank, "_ID")] = as.numeric(return_df$taxonID)
    return_df$OG_ID = return_df$taxonID
    return_df$taxonID = return_df$parentID
    return_df <- return_df %>%
      select(-parentID, -Rank, -name)
    #Merge matches and returned results
    return_df <- rbind(return_df, mismatches) 
  }
  
  #If matches is empty, skip it
  else{
    return_df = mismatches }
  
  #Return matches++
  return(return_df)
  }


# test <- Nodes %>%
Big_taxon_table <- Nodes %>%
  filter(Rank=="species") %>%
  # filter(taxonID  %in% c(1,2, 9606)) %>%
  select(taxonID) %>%
  mutate(OG_ID = taxonID) %>%
  do(recursive_taxon3(., Nodes, Names, "species")) %>%
  do(recursive_taxon3(., Nodes, Names, "genus")) %>%
  do(recursive_taxon3(., Nodes, Names, "family")) %>%
  do(recursive_taxon3(., Nodes, Names, "order")) %>%
  do(recursive_taxon3(., Nodes, Names, "class")) %>%
  do(recursive_taxon3(., Nodes, Names, "phylum")) %>%
  do(recursive_taxon3(., Nodes, Names, "kingdom")) %>%
  do(recursive_taxon3(., Nodes, Names, "superkingdom")) %>%
  do(recursive_taxon3(., Nodes, Names, "no_rank")) %>%
  filter(OG_ID !=1) %>%
  select(-taxonID, - OG_ID)

# write.table(filter(Big_taxon_table, no_rank=="cellular organisms"), file='Taxonomy_table.csv',
#             quote=FALSE, sep=',', col.names = NA)

```


##Read in species info for rRNAs
```{r}
 
bacteria_rRNA_species <- read.csv("../reference_sequences/rfam_bacteria_species.txt", sep="\t",
                                   col.names = c("gene", "genus", "species"))

#Merge species with genus and species
asdf <- rRNA_microbe_data %>%
  full_join(., bacteria_rRNA_species, by=c("gene")) %>%
  filter(!is.na(base))

# asdf <- Big_taxon_table %>%
#   select(-species, -species_ID, ) %>%
#   filter(!duplicated(genus_ID)) %>%
#   full_join(.,bacteria_rRNA_species, by=c("genus")) %>%
#   filter(!is.na(gene))

```


##Species check
```{r}

temp <- asdf %>%
  filter(subunit=="S",
         # genus=="Streptococcus",
         gene =="SSU_FP929046.1/591539-590690",
         treatment != "plusCMC",
         # gene =="SSU_CP000023.1/17818-19365",
         position >= 460,
         position <= 520
         )

ggplot(filter(temp
              ),
       aes(x=position, y=deletion / pileup, color=treatment)) +
  geom_line() +
  geom_text(aes(label=base, y=0), color="black") +
  facet_grid(cols=vars(paste0(genus," ", species)))

```


```{r}

temp <- asdf %>%
  filter(subunit =="S",
         position >= 400,
         position <= 600,
         position == 500,
         pileup >= 100,
         treatment !="plusCMC") %>%
  select(gene, position, pileup, deletion, treatment, genus, species) %>%
  pivot_wider(values_from = c("pileup","deletion", ), names_from = c("treatment"))

```

#Output bacteria stuff for Tao


```{r}
write_data_to_excel <- function(row){
  PATH = paste0("/Users/ckatanski/4SR/data/20210416_pseudoU_rRNA/source/", row$treatment,"_", row$subunit, "_", row$genus, ".csv")
  print(PATH)
  write.table(row$data, file=PATH, sep=",", row.names = F)
  return(data.frame(A=c(PATH)))
}



temp <- asdf %>%
  filter(gene %in% c("LSU_FP929046.1/589785-586950", "SSU_FP929046.1/591539-590690"),
         # position >= 400,
         # position <= 600,
         # position == 500,
         # pileup >= 100,
         # treatment !="plusCMC"
         ) %>%
  # select(gene, position, pileup, deletion, treatment, genus, species) %>%
  ungroup() %>%
  nest(data = c(-genus, -subunit, -treatment)) %>%
  group_by(genus, subunit, treatment) %>%
  do(., write_data_to_excel(.))
  # pivot_wider(values_from = c("pileup","deletion", ), names_from = c("treatment"))

```





#_
#Paper figures
##NGS TGS comparison
###Bacteria
```{r}
bacteria_TGS <- read.csv("../TGS-data/LSU_FP929046.1_selected_features_and_labels.csv", header=T) %>%
  mutate(treatment = "None",
         technique ="TGS") %>%
  select(-aligned_gene, -base_type) %>%
  filter(coverage >= 10) %>%
  pivot_longer(c(-treatment, -technique, -position), names_to = c("parameter"), values_to = c("value"))
  

temp <- asdf %>%
  filter(subunit=="L",
         # genus=="Streptococcus",
         gene =="LSU_FP929046.1/589785-586950",
         # grepl("LSU_FP929046.1", gene),
         treatment != "plusCMC",
         # gene =="SSU_CP000023.1/17818-19365",
         # position >= 460,
         # position <= 520
         ) %>%
  filter(pileup >= 100) %>%
  select(-A, -C, -G, -"T", -N, -insertion, -genus, -species) %>%
  mutate(technique = "NGS") %>%
  select(-gene, -base, -subunit) %>%
  mutate(del_site = deletion / pileup) %>%
  select(position, treatment, technique, del_site) %>%
  pivot_longer(c(-position, -treatment, -technique), names_to = c("parameter"), values_to = c("value")) 
temp$treatment <- recode(temp$treatment, "minusCMC"="None", "bisulfite"="bisulfite")
  

temp2 <- rbind(temp, bacteria_TGS)
temp2$parameter <- recode(temp2$parameter, 
                          "del_site"="Deletion\nrate", 
                          "mis"="Mutation\nrate", 
                          "modification_state"="Modification\nState",
                          "misC"= "Mutation\nrate to C")
temp2$treatment <- factor(temp2$treatment, levels = c("None", "bisulfite") )
temp2$technique <- recode(temp2$technique, "NGS"="Illumina", "TGS"="Nanopore")


plot1 <- ggplot(filter(temp2,
              parameter %nin% c("coverage", "Modification\nState", NA),
              ),
       aes(x=position, y=value, color=technique)) +
  geom_line() +
  ylab("Rate") +
  xlab("Position (nt)") +
  coord_cartesian(xlim = c(2480, 2580) ) +
  scale_y_continuous(limits=c(0,1), breaks=c(.25, 0.75)) +
  scale_color_manual(values = c("Illumina"="black", "Nanopore"="red")) +
  labs(color="Platform") +
  facet_nested(technique + treatment + parameter ~  "Faecalibacterium prausnitzii LSU") +
  theme(strip.background = element_rect(fill="white"),
        strip.text.y = element_text(angle=0) )

plot1

#Save the figure
layout <- rbind(c(1))
figure <- grid.arrange(plot1,  layout_matrix = layout)
path="../figures/"
figure_name="NGS_TGS_traces_bacteria"
width=5
height=3
scale=1.5
dpi=600

# ggsave(paste0(path, figure_name,".svg"), 
#        plot = figure,
#        scale = scale,
#        dpi = dpi,
#        width = width, 
#        height = height)

ggsave(paste0(path, figure_name,".png"),
       plot = figure,
       scale = scale,
       dpi = dpi,
       width = width,
       height = height)

```

###Human
```{r}
human_TGS <- read.csv("../TGS-data/human_18S_rRNA_features_and_labels.csv", header=T) %>%
  mutate(treatment = "None",
         technique ="TGS") %>%
  select(-aligned_gene, -base_type) %>%
  filter(coverage >= 10) %>%
  pivot_longer(c(-treatment, -technique, -position), names_to = c("parameter"), values_to = c("value"))
 

temp <- data_rRNA_assorted %>%
  filter(#subunit=="L",
         # genus=="Streptococcus",
         gene =="Homo_sapiens_18S_ribosomal",
         # grepl("LSU_FP929046.1", gene),
         treatment != "plusCMC",
         # gene =="SSU_CP000023.1/17818-19365",
         # position >= 460,
         # position <= 520
         ) %>%
  filter(pileup >= 10) %>%
  select(-A, -C, -G, -"T", -N, -insertion, -sample, -bin_start, -bin_stop) %>%
  mutate(technique = "NGS") %>%
  select(-gene, -base) %>%
  mutate(del_site = deletion / pileup) %>%
  select(position, treatment, technique, del_site) %>%
  pivot_longer(c(-position, -treatment, -technique), names_to = c("parameter"), values_to = c("value")) 
temp$treatment <- recode(temp$treatment, "minusCMC"="None", "bisulfite"="bisulfite")
  

temp2 <- rbind(temp, human_TGS)
temp2$parameter <- recode(temp2$parameter, 
                          "del_site"="Deletion\nrate", 
                          "mis"="Mutation\nrate", 
                          "modification_state"="Modification\nState",
                          "misC"= "Mutation\nrate to C")
temp2$treatment <- factor(temp2$treatment, levels = c("None", "bisulfite") )
temp2$technique <- recode(temp2$technique, "NGS"="Illumina", "TGS"="Nanopore")


plot1 <- ggplot(filter(temp2,
              parameter %nin% c("coverage", "Modification\nState",NA),
              ),
       aes(x=position, y=value, color=technique)) +
  geom_line() +
  ylab("Rate") +
  xlab("Position (nt)") +
  coord_cartesian(xlim = c(950,1100) ) +
  scale_y_continuous(limits=c(0,1), breaks=c(.25, 0.75)) +
  scale_color_manual(values = c("Illumina"="black", "Nanopore"="red")) +
  labs(color="Platform") +
  facet_nested(technique + treatment + parameter ~ "Human 18S") +
  theme(strip.background = element_rect(fill="white"),
        strip.text.y = element_text(angle=0) )

#Save the figure
layout <- rbind(c(1))
figure <- grid.arrange(plot1,  layout_matrix = layout)
path="../figures/"
figure_name="NGS_TGS_traces_human"
width=5
height=3
scale=1.5
dpi=600

# ggsave(paste0(path, figure_name,".svg"), 
#        plot = figure,
#        scale = scale,
#        dpi = dpi,
#        width = width, 
#        height = height)

ggsave(paste0(path, figure_name,".png"),
       plot = figure,
       scale = scale,
       dpi = dpi,
       width = width,
       height = height)

```



#_
#PKR-clip KIM 20210708
##Read in data
```{r}
PATH="../Kim_PKR/6_sam_counter/human_combine/"
FILES <-data.frame( file_name = list.files(PATH) ) %>%
  filter(!grepl("unassigned", file_name),
         file_name %nin% c("kallisto.out"))

#Split generic file names into salient parts
FILES <- FILES %>%
  separate(file_name, sep="_", c("junk1","library"), fill="right", remove=F) %>%
  separate(library, sep=".tsv", c("library", "junk2"), remove=T) %>%
  select(-junk1, -junk2) 


#PROJECT SPECIFIC DECODING
FILES$library <- recode(FILES$library, 
                          "SRR6456378"="input_M-phase",
                          "SRR6456379"="input_S-phase",
                          "SRR6456380"="PKR_M-phase",
                          "SRR6456381"="PKR_S-phase",
                          "SRR6456382"="J2_M-phase",
                          "SRR6456383"="J2_S-phase"
                          )
FILES <- FILES %>%
  separate(library, sep="_", c("material","phase"))



read_in_one <- function(row){
  output <- read.csv(paste0(PATH, row$file_name), header=T, sep="\t") %>%
    mutate(material = row$material,
           phase = row$phase,)
  #specify data types since enpty dataframes get confused and sad
  output$name <- as.character(output$name)
  output <- output %>%
    mutate(gene=name) %>%
    select(-name)
  return(output)
}

PKR_counts_data <- FILES %>%
  group_by(file_name) %>%
  do(read_in_one(.)) %>%
  ungroup() %>%
  select(-file_name)
```

##Gene names dict
```{r}
gene_id_dict <- read.csv("./human_combine_name_dict.txt",sep="\t")

#add names, combine isodecoders
temp <- PKR_counts_data %>%
  full_join(.,gene_id_dict, by=c("gene")) %>%
  group_by(gene_symbol, material, phase, gene_biotype) %>%
  summarise(count = sum(count))

#Normalize to rpm
PKR_counts <- temp %>%
  group_by(material, phase) %>%
  mutate(rpm = count / sum(count)) 
```

##Plot density
```{r}

big_increase_IFNgamma <- c("ODF3B","GBP2","GBP1","IFI6","SECTM1","IFI35","IFITM1",
                          "PSMB8","STAT1","WARS1","PSME1","LAP3","IL32","RNU1")

big_increase_IFNbeta <- c("ODF3B","IFI6","ISG15","IFI16","ISG15","IFI16","SECTM1",
                          "IFI35","OAS1","IFITM1","SAMD9","NT5C3A","STAT1","WARS1","PLSCR1")

#pivot wider by material
temp <- PKR_counts %>%
  select(-count) %>%
  filter(!is.na(material)) %>%
  pivot_wider(names_from=c("material"), values_from = c("rpm"))

ggplot(filter(temp),
       aes(x=PKR/input)) +
  geom_density() +
  geom_density(data=filter(temp, gene_symbol %in% big_increase_IFNbeta), color="red") +
  scale_x_log10() +
  facet_grid(rows=vars(phase))

```

##Plot box plot
```{r}

big_increase_IFNgamma <- c("ODF3B","GBP2","GBP1","IFI6","SECTM1","IFI35","IFITM1",
                          "PSMB8","STAT1","WARS1","PSME1","LAP3","IL32","RNU1")

big_increase_IFNbeta <- c("ODF3B","IFI6","ISG15","IFI16","ISG15","IFI16","SECTM1",
                          "IFI35","OAS1","IFITM1","SAMD9","NT5C3A","STAT1","WARS1","PLSCR1")

ABUNDANCE_FILTER = 20
#pivot wider by material
temp <- PKR_counts %>%
  filter(count >= ABUNDANCE_FILTER) %>%
  filter(gene_biotype =="protein_coding") %>%
  select(-count) %>%
  filter(!is.na(material)) %>%
  pivot_wider(names_from=c("material"), values_from = c("rpm")) %>%
  mutate(plot_group="all")

temp <- temp %>%
  filter(gene_symbol %in% big_increase_IFNgamma) %>%
  mutate(plot_group = "IFNgamma") %>%
  rbind(temp)
temp <- temp %>%
  filter(gene_symbol %in% big_increase_IFNbeta) %>%
  mutate(plot_group = "IFNbeta") %>%
  rbind(temp)

my_comparisons <- list( c("all", "IFNbeta"),
                        c("all", "IFNgamma")
                        )

ggplot(filter(temp),
       aes(y=PKR/input, x=plot_group)) +
  geom_boxplot(outlier.shape = NA
               ) +
  geom_jitter(width=0.1) +
  scale_y_log10() +
  geom_text_repel(data=filter(temp, plot_group=="IFNgamma"), aes(label=gene_symbol)) +
  stat_compare_means(comparisons = my_comparisons, size=4,  method="wilcox.test") +
  facet_grid(rows=vars(phase)) +
  theme(strip.background = element_blank(),
        strip.text.y = element_text(angle=0)) +
  ggtitle(paste0("Abundnace filter: >",ABUNDANCE_FILTER, " reads"))

```


#Read in TGS expression from Sihao
```{r}
#Read in Sihao's expression data
TGS_expression_data <- read.csv("../TGS-data/all_gene_expression_level.csv", header=T)

temp <- TGS_expression_data %>%
  pivot_longer(c(-gene_name, -gene_type, -gene_ID), names_to = c("treatment"), values_to = c("count")) %>%
  group_by(gene_name, treatment, gene_type) %>%
  summarise(count = sum(count)) %>%
  group_by(treatment) %>%
  mutate(rpm = count / sum(count))
  
EXPRESSION_FILTER = 5

temp <- temp %>%
  filter(count >= EXPRESSION_FILTER)
TGS_expression <- temp %>%
  filter(treatment == "Untreated_express") %>%
  mutate(untreated = rpm) %>%
  ungroup() %>%
  select(-treatment, -count, -rpm) %>%
  full_join(temp,., by=c("gene_name","gene_type")) %>%
  mutate(fold_change = rpm / untreated) %>%
  filter(treatment != "Untreated_express") %>%
  mutate(gene_symbol = gene_name,
         gene_biotype = gene_type) %>%
  select(-gene_name, gene_type)

```

##Combine PKR and TGS
```{r}

temp <- PKR_counts %>%
  filter(material=="input") %>%
  ungroup() %>%
  mutate(input = rpm) %>%
  select(-material, -count, -rpm) %>%
  full_join(PKR_counts, ., by=c("gene_symbol","gene_biotype","phase")) %>%
  filter(material != "input") 

COMBINE <- TGS_expression %>%
  select(-count, -rpm) %>%
  full_join(., temp, by=c("gene_symbol","gene_biotype")) %>%
  filter(!is.na(material),
        !is.na(phase),
        !is.na(treatment))
```

##Scatter plot!
```{r}
temp <- COMBINE %>%
  filter(gene_biotype=="protein_coding")

ggplot(filter(temp,
              # material=="PKR", 
              # treatment =="IFN_gamma_express"
              ),
       aes(x=fold_change, y=rpm/input)) +
  geom_point() +
  facet_grid(rows=vars(phase, treatment),
             cols=vars(material)) +
  scale_y_log10() +
  scale_x_log10() +
  stat_smooth(data=filter(temp, fold_change >= 2, #treatment=="IFN_gamma_express"
                          ),method="lm")

```

##Boxplots
```{r}
temp <- COMBINE %>%
  mutate(induction_group = ifelse(fold_change >= 2, "induced", "other")) %>%
  filter(material=="PKR")

my_comparisons <- list( c("other", "induced")
                        )

ggplot(filter(temp,
              # treatment =="IFN_gamma_express",
              # material =="PKR"
              ),
       aes(x=induction_group, y=rpm / input)) +
  geom_boxplot() +
  scale_y_log10() +
  facet_grid(cols=vars(treatment),
             rows=vars(phase)) +
  stat_compare_means(comparisons = my_comparisons, size=4,  method="wilcox.test") 

asdf <- filter(temp, induction_group =="induced", treatment =="IFN_gamma_express")

```

##Boxplot, 2 groups
```{r}

INDUCTION_THRESHOLD = 2

TGS_detected <- filter(TGS_expression, treatment =="IFN_gamma_express", fold_change >= 0)$gene_symbol
big_increase_IFNgamma <- filter(TGS_expression, treatment =="IFN_gamma_express", fold_change >= INDUCTION_THRESHOLD)$gene_symbol
big_increase_IFNbeta <- filter(TGS_expression, treatment =="IFN_beta_express", fold_change >= INDUCTION_THRESHOLD)$gene_symbol


ABUNDANCE_FILTER = 20
#pivot wider by material
temp <- PKR_counts %>%
  filter(count >= ABUNDANCE_FILTER) %>%
  # filter(gene_symbol %in% TGS_detected) %>%
  filter(gene_biotype == "protein_coding") %>%
  select(-count) %>%
  filter(!is.na(material)) %>%
  pivot_wider(names_from=c("material"), values_from = c("rpm")) %>%
  mutate(plot_group="all")

temp <- temp %>%
  filter(gene_symbol %in% big_increase_IFNgamma) %>%
  mutate(plot_group = "IFNgamma") %>%
  rbind(temp)
temp <- temp %>%
  filter(gene_symbol %in% big_increase_IFNbeta) %>%
  mutate(plot_group = "IFNbeta") %>%
  rbind(temp)

my_comparisons <- list( c("all", "IFNbeta"),
                        c("all", "IFNgamma")
                        )

ggplot(filter(temp),
       aes(y=PKR/input, x=plot_group)) +
  geom_boxplot(outlier.shape = NA
               ) +
  geom_jitter(width=0.1) +
  scale_y_log10() +
  # geom_text_repel(data=filter(temp, plot_group=="IFNgamma"), aes(label=gene_symbol)) +
  stat_compare_means(comparisons = my_comparisons, size=4,  method="wilcox.test") +
  facet_grid(rows=vars(phase)) +
  theme(strip.background = element_blank(),
        strip.text.y = element_text(angle=0)) +
  ggtitle(paste0("Abundnace filter: >",ABUNDANCE_FILTER, " reads","\n","Inudcted threshold: >",INDUCTION_THRESHOLD,"-fold"))

```


##Box_plot many groups
```{r}
INDUCTION_THRESHOLD = 10

TGS_detected <- filter(TGS_expression, treatment =="IFN_gamma_express", fold_change >= 0)$gene_symbol
big_increase_IFNgamma <- filter(TGS_expression, treatment =="IFN_gamma_express", fold_change >= INDUCTION_THRESHOLD)$gene_symbol
big_increase_IFNbeta <- filter(TGS_expression, treatment =="IFN_beta_express", fold_change >= INDUCTION_THRESHOLD)$gene_symbol


induced_bins=c(.5, 1, 2, 300)
IFNgamma_threshold <- TGS_expression %>%
  filter(treatment=="IFN_gamma_express") %>%
  ungroup() %>%
  mutate(induced_bin = ifelse(fold_change < induced_bins[1], paste0("<",induced_bins[1]),
                              ifelse(fold_change >= induced_bins[1] & fold_change < induced_bins[2],
                                     paste0(induced_bins[1],"-",induced_bins[2]),
                                     ifelse(fold_change >= induced_bins[2] & fold_change < induced_bins[3],
                                            paste0(induced_bins[2],"-",induced_bins[3]),
                                            ifelse(fold_change >= induced_bins[3] & fold_change < induced_bins[4], 
                                                   paste0(induced_bins[3],"-",induced_bins[4]),
                                                   ifelse(fold_change >= induced_bins[4] , 
                                                          paste0(">",induced_bins[4]), "OTHER")))))) %>%
  select(treatment, gene_symbol, induced_bin)

IFNbeta_threshold <- TGS_expression %>%
  filter(treatment=="IFN_beta_express") %>%
  ungroup() %>%
  mutate(induced_bin = ifelse(fold_change < induced_bins[1], paste0("<",induced_bins[1]),
                              ifelse(fold_change >= induced_bins[1] & fold_change < induced_bins[2],
                                     paste0(induced_bins[1],"-",induced_bins[2]),
                                     ifelse(fold_change >= induced_bins[2] & fold_change < induced_bins[3],
                                            paste0(induced_bins[2],"-",induced_bins[3]),
                                            ifelse(fold_change >= induced_bins[3] & fold_change < induced_bins[4], 
                                                   paste0(induced_bins[3],"-",induced_bins[4]),
                                                   ifelse(fold_change >= induced_bins[4] , 
                                                          paste0(">",induced_bins[4]), "OTHER")))))) %>%
  select(treatment, gene_symbol, induced_bin)


ABUNDANCE_FILTER = 1
#pivot wider by material
temp <- PKR_counts %>%
  filter(count >= ABUNDANCE_FILTER) %>%
  filter(gene_symbol %in% TGS_detected) %>%
  filter(gene_biotype == "protein_coding") %>%
  select(-count) %>%
  filter(!is.na(material)) %>%
  pivot_wider(names_from=c("material"), values_from = c("rpm"))

temp1 <- temp %>%
  filter(gene_symbol %in% IFNgamma_threshold$gene_symbol) %>%
  full_join(., IFNgamma_threshold, by=c("gene_symbol"))
temp2 <- temp %>%
  filter(gene_symbol %in% IFNbeta_threshold$gene_symbol) %>%
  full_join(., IFNbeta_threshold, by=c("gene_symbol")) 

temp3 <- temp %>%
  mutate(induced_bin = "all",
         treatment ="IFN_beta_express")
temp4 <- temp %>%
  mutate(induced_bin = "all",
         treatment ="IFN_gamma_express")

temp <- rbind(temp1, temp2, temp3, temp4) %>%
  filter(!is.na(phase))

temp$induced_bin <- factor(temp$induced_bin, levels=c("<0.5","0.5-1", "1-2", "2-300", ">100","all"))

my_comparisons <- list( c("<0.5", "2-300")
                        )

ggplot(filter(temp),
       aes(y=PKR/input, x=induced_bin)) +
  geom_boxplot(outlier.shape = NA
               ) +
  geom_jitter(width=0.1) +
  scale_y_log10() +
  # geom_text_repel(data=filter(temp, plot_group=="IFNgamma"), aes(label=gene_symbol)) +
  stat_compare_means(comparisons = my_comparisons, size=4,  method="wilcox.test") +
  facet_grid(rows=vars(phase),
             cols=vars(treatment)) +
  theme(strip.background = element_blank(),
        strip.text.y = element_text(angle=0)) +
  ggtitle(paste0("Abundnace filter: >",ABUNDANCE_FILTER, " reads","\n","Inudcted threshold: >",INDUCTION_THRESHOLD,"-fold"))

```

#_
#Single read
##WARS1
###Read in data
```{r}
WARS1_data1 <- read.csv("../TGS-data/WARS1_SingleRead/all_U_prediction_WARS1_u.csv", header=F, sep=",", col.names = c("read","gene","position","base","prob_U","prob_psU")) %>%
  select(-prob_U) %>%
  mutate(treatment ="untreated")

WARS1_data2 <- read.csv("../TGS-data/WARS1_SingleRead/all_U_prediction_WARS1_g.csv", header=F, sep=",", col.names = c("read","gene","position","base","prob_U","prob_psU")) %>%
  select(-prob_U) %>%
  mutate(treatment ="gamma")

WARS1_data3 <- read.csv("../TGS-data/WARS1_SingleRead/all_U_prediction_WARS1_b.csv", header=F, sep=",", col.names = c("read","gene","position","base","prob_U","prob_psU")) %>%
  select(-prob_U) %>%
  mutate(treatment ="beta")

WARS1_data <- rbind(WARS1_data1,WARS1_data2, WARS1_data3)

```

###Calculate number_psU per read
```{r}
temp <- WARS1_data %>%
  group_by(read, gene, treatment) %>%
  mutate(counter = 1) %>%
  filter(prob_psU >= 0.95) %>%
  summarise(num_psU = sum(counter)) %>%
  group_by(treatment)

ggplot(temp, aes(x=num_psU, color=treatment)) +
  # geom_histogram(aes(y = stat(count / sum(count))), alpha=0.4, position = position_dodge())
  geom_density(bin_width=1)

```

```{r}
temp <- WARS1_data %>%
  group_by(read, gene, treatment) %>%
  mutate(counter = 1) %>%
  filter(prob_psU >= 0.80) %>%
  summarise(num_psU = sum(counter)) %>%
  group_by(treatment)


temp <- temp %>%
  group_by(treatment) %>%
  mutate(total_reads = length(unique(read))) %>%
  group_by(treatment, num_psU) %>%
  mutate(num_reads = length(unique(read))) %>%
  ungroup()


ggplot(temp, aes(x=num_psU, y=num_reads / total_reads, color=treatment)) +
  geom_line() +
  geom_point() +
  ylab("fraction of reads") +
  xlab("number psU sites per read") +
  facet_grid(rows=vars(treatment))

```


###psU RATE
```{r}

temp <- WARS1_data %>%
  group_by(read, gene, treatment) %>%
  summarise(read_length = length(unique(position)))

ggplot(temp, aes(x=read_length, color=treatment)) +
  geom_density()

temp <- WARS1_data %>%
  group_by(read, gene, treatment) %>%
  mutate(counter = 1,
         read_length = 4 * length(unique(position))) %>%
  filter(prob_psU >= 0.85) %>%
  summarise(rate_psU = sum(counter)/median(read_length),
            ) %>%
  group_by(treatment)

# temp <- temp %>%
#   group_by(treatment) %>%
#   mutate(total_reads = length(unique(read))) %>%
#   group_by(treatment, num_psU) %>%
#   mutate(num_reads = length(unique(read))) %>%
#   ungroup()


ggplot(temp, aes(x=rate_psU, color=treatment)) +
  # stat_ecdf() +
  geom_density()

my_comparisons <- list( c("untreated", "beta"),
                        c("beta", "gamma"),
                        c("untreated", "gamma")
                        )

temp$treatment <- factor(temp$treatment, levels=c("untreated","beta","gamma"))

ggplot(temp, aes(x=treatment, y=rate_psU)) +
  geom_boxplot() +
  stat_compare_means(comparisons = my_comparisons, size=4,  method="wilcox.test") 

asdf <- temp %>%
  group_by(treatment, gene) %>%
  summarise(mean_rate = mean(rate_psU),
            num_reads = length(unique(read)))

```


###count psU in long reads
```{r}
temp <- WARS1_data %>%
  group_by(read, gene, treatment) %>%
  mutate(read_length = length(unique(position))) %>%
  filter(read_length >= 500) %>%
  mutate(counter = 1) %>%
  filter(prob_psU >= 0.85) %>%
  summarise(num_psU = sum(counter)) %>%
  group_by(treatment)

ggplot(temp, aes(num_psU, color=treatment)) +
  stat_ecdf()

ggplot(temp, aes(x=treatment, y=num_psU)) +
  geom_boxplot() +
  stat_compare_means(comparisons = my_comparisons, size=4,  method="wilcox.test") 

temp <- temp %>%
  group_by(treatment) %>%
  mutate(total_reads = length(unique(read))) %>%
  group_by(treatment, num_psU) %>%
  mutate(num_reads = length(unique(read))) %>%
  ungroup()


ggplot(filter(temp, 
              treatment !="untreated"
              ), aes(x=num_psU, y=num_reads / total_reads, color=treatment)) +
  geom_line() +
  geom_point() +
  ylab("fraction of reads") +
  xlab("number psU sites per read") 
  # facet_grid(rows=vars(treatment))

```

###psU RATE in long reads
```{r}

temp <- WARS1_data %>%
  group_by(read, gene, treatment) %>%
  summarise(read_length = length(unique(position)))

ggplot(temp, aes(x=read_length, color=treatment)) +
  geom_density()

temp <- WARS1_data %>%
  group_by(read, gene, treatment) %>%
  mutate(read_length = length(unique(position))) %>%
  filter(read_length >= 500) %>%
  group_by(read, gene, treatment) %>%
  mutate(counter = 1,
         read_length = 4 * length(unique(position))) %>%
  filter(prob_psU >= 0.9) %>%
  summarise(rate_psU = sum(counter)/median(read_length),
            ) %>%
  group_by(treatment)

# temp <- temp %>%
#   group_by(treatment) %>%
#   mutate(total_reads = length(unique(read))) %>%
#   group_by(treatment, num_psU) %>%
#   mutate(num_reads = length(unique(read))) %>%
#   ungroup()


ggplot(temp, aes(x=rate_psU, color=treatment)) +
  # stat_ecdf() +
  geom_density()

my_comparisons <- list( c("untreated", "beta"),
                        c("beta", "gamma"),
                        c("untreated", "gamma")
                        )

temp$treatment <- factor(temp$treatment, levels=c("untreated","beta","gamma"))

ggplot(temp, aes(x=treatment, y=rate_psU)) +
  geom_boxplot() +
  stat_compare_means(comparisons = my_comparisons, size=4,  method="wilcox.test") 

asdf <- temp %>%
  group_by(treatment, gene) %>%
  summarise(mean_rate = mean(rate_psU),
            num_reads = length(unique(read)))

```

##psU regions?
```{r}
temp <- WARS1_data %>%
  group_by(treatment, position, gene) %>%
  summarise(mean_psU_prob = median(prob_psU))

ggplot(temp, aes(x=position, y=mean_psU_prob, color=treatment)) +
  geom_point(alpha=0.2) +
  coord_cartesian(xlim = c(6660000, 6715000))


```


##IFITM3 single reaad
```{r}
IFITM3_data1 <- read.csv("../TGS-data/IFITM3_single_read/all_U_prediction_IFITM3_u.csv", header=F, sep=",", col.names = c("read","gene","position","base","prob_U","prob_psU")) %>%
  select(-prob_U) %>%
  mutate(treatment ="untreated")

IFITM3_data2 <- read.csv("../TGS-data/IFITM3_single_read/all_U_prediction_IFITM3_g.csv", header=F, sep=",", col.names = c("read","gene","position","base","prob_U","prob_psU")) %>%
  select(-prob_U) %>%
  mutate(treatment ="gamma")

IFITM3_data3 <- read.csv("../TGS-data/IFITM3_single_read/all_U_prediction_IFITM3_b.csv", header=F, sep=",", col.names = c("read","gene","position","base","prob_U","prob_psU")) %>%
  select(-prob_U) %>%
  mutate(treatment ="beta")

IFITM3_data <- rbind(IFITM3_data1,IFITM3_data2, IFITM3_data3)

```

###psU rate
```{r}

temp <- IFITM3_data %>%
  group_by(read, gene, treatment) %>%
  summarise(read_length = length(unique(position)))

ggplot(temp, aes(x=read_length, color=treatment)) +
  geom_density()

temp <- IFITM3_data %>%
  group_by(read, gene, treatment) %>%
  mutate(counter = 1,
         read_length = 4 * length(unique(position))) %>%
  filter(prob_psU >= 0.85) %>%
  summarise(rate_psU = sum(counter)/median(read_length),
            ) %>%
  group_by(treatment)

# temp <- temp %>%
#   group_by(treatment) %>%
#   mutate(total_reads = length(unique(read))) %>%
#   group_by(treatment, num_psU) %>%
#   mutate(num_reads = length(unique(read))) %>%
#   ungroup()


ggplot(temp, aes(x=rate_psU, color=treatment)) +
  stat_ecdf() 
  # geom_density()

my_comparisons <- list( c("untreated", "beta"),
                        c("beta", "gamma"),
                        c("untreated", "gamma")
                        )

temp$treatment <- factor(temp$treatment, levels=c("untreated","beta","gamma"))

ggplot(temp, aes(x=treatment, y=rate_psU)) +
  geom_boxplot() +
  stat_compare_means(comparisons = my_comparisons, size=4,  method="wilcox.test") 

asdf <- temp %>%
  group_by(treatment, gene) %>%
  summarise(mean_rate = mean(rate_psU),
            num_reads = length(unique(read)))

```

##Read in single reads
```{r}
#Noah, you might have to change this folder name
PATH = "../TGS-data/single_read/"
FILES <-data.frame( file_name = list.files(PATH) )

#Split generic file names into salient parts
FILES <- FILES %>%
  extract(file_name, "(.*).csv", into="file_name2", remove=FALSE) %>% #"extract" is used for 'regular expressions'
  separate(file_name2, remove=TRUE, sep="_", c("junk1","junk2","junk3","gene", "treatment") ) %>% 
  select(-junk1, -junk2, -junk3) 

#A function that reads in a data file and adds above annotation
read_in_one <- function(row){
  output <- read.csv(paste0(PATH, row$file_name), header=F, sep=",",
                     col.names = c("read","gene","position","base","prob_U","prob_psU") ) %>%
    mutate(gene = row$gene,
           treatment  = row$treatment) %>%
    select(-prob_U, -base)
  return(output)
}


#Up to this point the "FILES" data frame is just identifying info, no actual tRNA data.
#The last step here is to run the funciton above such that it reads in the data in file with 
#"file_name" and identifies all of that data with the right patient, tissue, DM, bin_start, etc data

#A loop to read in all the data files and add the annotation
TGS_single_read <- FILES %>%
  group_by(file_name) %>%
  # filter(grepl("IFIT", file_name)) %>%
  do(read_in_one(.)) %>%
  ungroup()

TGS_single_read$treatment <- recode(TGS_single_read$treatment, "u"="untreated", "b"="beta", "g"="gamma")

```

###psU RATE
```{r}

# temp <- TGS_single_read %>%
#   group_by(read, gene, treatment) %>%
#   summarise(read_length = length(unique(position)))
# 
# ggplot(temp, aes(x=read_length, color=treatment)) +
#   geom_density() +
#   facet_wrap(~gene)

THRESHOLD = 0.95

temp <- TGS_single_read %>%
  group_by(read, gene, treatment) %>%
  mutate(counter = 1,
         read_length = 4 * length(unique(position))) %>%
  filter(prob_psU >= THRESHOLD) %>%
  summarise(rate_psU = sum(counter)/median(read_length),
            ) %>%
  group_by(treatment, gene)



my_comparisons <- list( c("untreated", "beta"),
                        c("beta", "gamma"),
                        c("untreated", "gamma")
                        )

temp$treatment <- factor(temp$treatment, levels=c("untreated","beta","gamma"))
temp <- temp %>%
  mutate(induction = gene)
temp$induction <- recode(temp$induction, "RPL5"="down","RPL7"="down","TGFBI"="down","IFITM3"="up","MT2A"="up","B2M"="up", "WARS1"="up")

ggplot(temp, aes(x=rate_psU, color=induction, group=interaction(gene, induction, treatment))) +
  stat_ecdf() +
  # geom_density() +
  facet_wrap(~treatment) +
  ggtitle(paste0("psU sites with >= ",THRESHOLD," probability")) 

ggplot(temp, aes(x=treatment, y=rate_psU)) +
  geom_boxplot() +
  stat_compare_means(comparisons = my_comparisons, size=4,  method="t.test") +
  facet_nested_wrap(induction ~ gene) +
  ggtitle(paste0("psU sites with >=",THRESHOLD," probability")) 
  # coord_cartesian(ylim = c(0,0.1))

asdf <- temp %>%
  group_by(induction, gene, treatment) %>%
  summarise(mean_rate = mean(rate_psU),
            num_reads = length(unique(read))) 


ggplot(filter(TGS_single_read, gene =="B2M", treatment=="beta",
              read == 0),
       aes(y=prob_psU, x=position, color=prob_psU, group=interaction(read, treatment, gene))) +
  geom_point() +
  geom_line()

```

##IFITM3
```{r}
temp1 <- TGS_single_read %>%
  group_by(gene, position, treatment) %>%
  mutate(count=1) %>%
  summarise(mean_psU = mean(prob_psU, na.rm=T) / var(prob_psU, na.rm=T),
            pileup = sum(count) ) %>%
  filter(pileup >=50) %>%
  mutate(induction = gene)
temp1$induction <- recode(temp1$induction, "RPL5"="down","RPL7"="down","TGFBI"="down","IFITM3"="up","MT2A"="up","B2M"="up", "WARS1"="up")
  
ggplot(filter(temp1, gene =="IFITM3",
              # induction =="up"
              ),
       aes(y=mean_psU, x=position, group=interaction(treatment, gene))) +
  geom_point() +
  geom_line() +
  # coord_cartesian(xlim = c(44716500, 44719000)) +
  facet_grid(rows=vars(treatment),
             cols=vars(gene)) +
  geom_vline(color="red", xintercept = c(134765801, 134766020 ,134766875))

asdf <- temp1 %>%
  filter(mean_psU >= 25,
         gene=="IFITM3",
         treatment=="untreated")

temp2 <- TGS_single_read %>%
  filter(gene =="IFITM3") %>%
  filter(
         ( (position > 134765801 - 5 & position < 134765801 + 5)| 
           (position > 134766020 - 5 & position < 134766020 + 5)|
           (position > 134766875 - 5 & position < 134766875 + 5) ) ) %>%
  mutate(readX = paste0(treatment, read)) 

temp2 <- temp2 %>%
  select(-file_name) %>%
  mutate(position = paste0("position", position)) %>%
  pivot_wider(names_from = c("position"), values_from = c("prob_psU"))

asdf <- temp2[order(temp2$position134766020),]

temp2 <- asdf %>%
  filter(position134765797 >=0 ) %>%
  pivot_longer(c(-read, -readX, -treatment, -gene), names_to = c("position"), values_to = c("prob_psU"))

temp2$position <- factor(temp2$position)

ggplot(temp2,
       aes(y=interaction(treatment, readX), x=position, fill=prob_psU)) +
  geom_tile() +
  theme(axis.text.y = element_blank(),
        axis.ticks.y = element_blank())
```

##B2M
```{r}
GENE = "B2M"

temp1 <- TGS_single_read %>%
  group_by(gene, position, treatment) %>%
  mutate(count=1) %>%
  summarise(mean_psU = mean(prob_psU, na.rm=T) ,
            var_psU = var(prob_psU, na.rm=T),
            pileup = sum(count) ) %>%
  filter(pileup >=100,
         # mean_psU >= 0.2
         ) %>%
  mutate(metric = var_psU / mean_psU )

  
ggplot(filter(temp1, gene == GENE,
              # metric > 0.05
              ),
       aes(y=mean_psU, x=position, group=interaction(treatment, gene))) +
  geom_point() +
  geom_line() +
  geom_point(data=filter(temp1, gene== GENE, 
                         metric >= 0.05
                         ), color="red") +
  # coord_cartesian(xlim = c(44716500, 44719000)) +
  facet_grid(rows=vars(treatment),
             cols=vars(gene)) 

asdf <- temp1 %>%
  filter(metric >= 0,
         gene== GENE,
         treatment=="untreated")

temp2 <- TGS_single_read %>%
  filter(gene == GENE) %>%
  # filter(
  #        ( (position > 134765801 - 5 & position < 134765801 + 5)| 
  #          (position > 134766020 - 5 & position < 134766020 + 5)|
  #          (position > 134766875 - 5 & position < 134766875 + 5) ) ) %>%
  filter(position %in% c(asdf$position)) %>%
  mutate(readX = paste0(treatment, read)) 

temp2 <- temp2 %>%
  select(-file_name) %>%
  mutate(position = paste0("position", position)) %>%
  pivot_wider(names_from = c("position"), values_from = c("prob_psU"))

# asdf <- temp2[order(temp2$position134766020),]

temp2 <- temp2 %>%
  # filter(position56609495 >=0 ) %>%
  pivot_longer(c(-read, -readX, -treatment, -gene), names_to = c("position"), values_to = c("prob_psU"))

temp2$position <- factor(temp2$position)

ggplot(filter(temp2),
       aes(y=interaction(treatment, readX), x=position, fill=prob_psU)) +
  geom_tile() +
  theme(axis.text.y = element_blank(),
        axis.ticks.y = element_blank()) 
  # facet_nested(. ~treatment)


```


##Fit bimodal
```{r}
GENE = "B2M"
PILEUP = 1000

temp1 <- TGS_single_read %>%
  group_by(gene, position, treatment) %>%
  mutate(count=1) %>%
  summarise(mean_psU = mean(prob_psU, na.rm=T) ,
            var_psU = var(prob_psU, na.rm=T),
            pileup = sum(count) ) %>%
  filter(pileup >= PILEUP,
         # gene == GENE
         ) %>%
  mutate(positionX = paste0(treatment, position))


temp2 <- TGS_single_read %>%
  mutate(positionX = paste0(treatment, position)) %>%
  group_by(gene, treatment) %>%
  mutate(sample_pileup = length(unique(read))) %>%
  filter(positionX %in% c(temp1$positionX)) %>%
  select(-positionX)

#A function for mixture modeling
misture_model <- function(df){
  x1 = length(df$prob_psU)
  x2 = median(df$sample_pileup)
  # print(x1)
  asdf <- normalmixEM(df$prob_psU, fast=T,
                      maxit = 50,
                      k=2, arbmean=T,
                      # lambda=c(0.9, 0.1),
                      # mu = c(0,1),
                      # arbvar=T,
                      verb=F
                      )
  
  output = data.frame(mu_difference = abs(c(asdf$mu[2] - asdf$mu[1])),
                      lambda_1 = asdf$lambda[1], 
                      lambda_2 = asdf$lambda[2],
                      coverage = x1 / x2
                      )
  return(output)
}


set.seed(123)
temp3 <- temp2 %>%
  group_by(treatment, position, gene) %>%
  do(misture_model(.))

ggplot(temp3, aes(x=coverage, color=treatment)) +
  geom_density()

ggplot(filter(temp3, gene=="B2M",
              lambda_1 >= 0.05 & lambda_2 >= 0.05
              ),
       aes(x=position, y=mu_difference, color =  abs(lambda_2 - lambda_1) )) +
  geom_point() +
  geom_hline(yintercept = c(0.5), linetype=2, alpha=0.3) +
  coord_cartesian(xlim = c(44717500, 44718300)) 



asdf <- filter(temp3, mu_difference >= 0.5, 
               lambda_1 >= 0.2 & lambda_2 >= 0.2,
               coverage >= 0.75
               )





temp4 <- temp2 %>%
  filter(#gene == GENE,
         # treatment == "untreated"
         ) %>%
  filter(position %in% asdf$position) %>%
  mutate(readX = paste0(treatment, read)) 

temp4 <- temp4 %>%
  select(-file_name) %>%
  mutate(position = paste0("position", position)) 


temp4$position <- factor(temp4$position)

# ggplot(filter(temp4),
#        aes(y=interaction(treatment, readX), x=position, fill=prob_psU)) +
#   geom_tile() +
#   scale_fill_gradient2(low="grey",  mid="grey", high="blue", midpoint = 0.05) +
#   theme(axis.text.y = element_blank(),
#         axis.ticks.y = element_blank(),
#         axis.text.x = element_text(angle=90)) 

temp5 <- temp4 %>%
  mutate(counter =1) %>%
  group_by(treatment, read, gene) %>%
  filter(prob_psU >= 0.5) %>%
  summarise(num_site = sum(counter))
  
ggplot(temp5, aes(x=num_site, color=treatment)) +
  stat_ecdf() +
  facet_wrap(~gene)

temp5$treatment <- factor(temp5$treatment, levels=c("untreated","beta","gamma"))
my_comparisons <- list( c("untreated", "beta"),
                        c("beta", "gamma"),
                        c("untreated", "gamma")
                        )

ggplot(temp5, aes(y=num_site, x=treatment))+
  # geom_jitter(width=0.2, height = 0) +
  # geom_boxplot(outlier.shape = NA,size=2, color="red") +
  geom_boxplot() +
  # geom_jitter() +
  stat_compare_means(comparisons = my_comparisons, size=4,  method="wilcox.test") +
  facet_wrap(~gene)

# wasdf <- temp5 %>%
#   group_by(treatment, gene) %>%
#   summarise(mean_num_site = mean(num_site))
# 
# wasdf2 <- temp4 %>%
#   group_by(gene) %>%
#   select(gene, treatment, position) %>%
#   filter( !duplicated(position)) %>%
#   separate(position, sep=c(8), into=c("junk", "position")) %>%
#   select(-junk)

# write.csv(wasdf2, col.names=T, sep=",", row.names = F,
#           file="./TGS_single_read_sites_with_2_populations.txt")
#   

```

##Distributions - plot and KS test
```{r}

# M2T
# 56609451
# 56609455

# IFITM3
# 134765832
# 134765835

tempA <- temp2 %>%
  filter(gene=="IFITM3",
         treatment=="untreated") %>%
  filter(position %in% c(134765832, 134765835)) %>%
  mutate(position = paste0("position", position)) %>%
  pivot_wider(names_from = c("position"), values_from = c("prob_psU")) %>%
  mutate(is_site_1 = "all")

tempA <- tempA %>%
  mutate(is_site_1 = ifelse(position134765832 >= 0.5, "high", "low")) %>%
  rbind(tempA) %>%
  filter(!is.na(is_site_1)) 

plot1 <- ggplot(tempA,
       aes(x=position134765835, color=is_site_1)) +
  stat_ecdf() +
  ylab("Fraction of reads") +
  xlab("Prob psU at\ndownstream site") +
  labs(color="prob psU at\nupstream site") +
  theme(legend.position = "None") +
  ggtitle("IFITM3")


tempB <- temp2 %>%
  filter(gene=="MT2A",
         treatment=="untreated") %>%
  filter(position %in% c(56609451, 56609455)) %>%
  mutate(position = paste0("position", position)) %>%
  pivot_wider(names_from = c("position"), values_from = c("prob_psU")) %>%
  mutate(is_site_1 = "all")

tempB <- tempB %>%
  mutate(is_site_1 = ifelse(position56609451 >= 0.5, "high", "low")) %>%
  rbind(tempB) %>%
  filter(!is.na(is_site_1))  %>%
  group_by(treatment, gene, is_site_1) %>%
  mutate(num_reads = length(unique(read)))


plot2 <- ggplot(tempB,
       aes(x=position56609455, color=is_site_1)) +
  stat_ecdf() +
  ylab("Fraction of reads") +
  xlab("Prob psU at\ndownstream site") +
  labs(color="prob psU at\nupstream site") +
  theme(axis.title.y = element_blank(),
        axis.text.y = element_blank()) +
  ggtitle("MT2A")

figure_legend <- get_legend(plot2+theme(legend.position = "bottom",
                                     legend.background = element_blank()) )
layout <- rbind(c(1,2), 
                c(1,2),
                c(1,2),
                c(3,3))
figure <- grid.arrange(plot1, plot2 + theme(legend.position = "none"), figure_legend,  layout_matrix = layout)

path="./"
figure_name="single_read_2_site_distributions"
width=4
height=4
scale=1.5
dpi=600
# ggsave(paste0(path, figure_name,".svg"), 
#        plot = figure,
#        scale = scale,
#        dpi = dpi,
#        width = width, 
#        height = height)

# ggsave(paste0(path, figure_name,".png"),
#        plot = figure,
#        scale = scale,
#        dpi = dpi,
#        width = width,
#        height = height)



# #https://rdrr.io/github/happyrabbit/DataScienceR/man/pairwise_ks_test.html
# install.packages("remotes")
# remotes::install_github("happyrabbit/DataScienceR")
# library(DataScienceR)
# 
# pairwise_ks_test(tempA$position134765835, tempA$is_site_1, warning = -1, alternative="two.sided")
# pairwise_ks_test(tempB$position56609455, tempB$is_site_1, warning = -1, alternative="two.sided")
# 

```

#_
Distributions for interferon treatment
```{r}

tempA <- temp2 %>%
  filter(gene=="IFITM3",
         # treatment=="untreated"
         ) %>%
  filter(position %in% c(134765832, 134765835)) %>%
  mutate(position = paste0("position", position)) %>%
  pivot_wider(names_from = c("position"), values_from = c("prob_psU")) %>%
  mutate(is_site_1 = "all")

tempA <- tempA %>%
  mutate(is_site_1 = ifelse(position134765832 >= 0.5, "high", "low")) %>%
  rbind(tempA) %>%
  filter(!is.na(is_site_1)) 

plot1 <- ggplot(tempA,
       aes(x=position134765835, color=treatment)) +
  stat_ecdf() +
  ylab("Fraction of reads") +
  xlab("Prob psU at\ndownstream site") +
  labs(color="prob psU at\nupstream site") +
  theme(legend.position = "None") +
  ggtitle("IFITM3") +
  facet_grid(rows=vars(is_site_1))


tempB <- temp2 %>%
  filter(gene=="MT2A",
         # treatment=="untreated"
         ) %>%
  filter(position %in% c(56609451, 56609455)) %>%
  mutate(position = paste0("position", position)) %>%
  pivot_wider(names_from = c("position"), values_from = c("prob_psU")) %>%
  mutate(is_site_1 = "all")

tempB <- tempB %>%
  mutate(is_site_1 = ifelse(position56609451 >= 0.5, "high", "low")) %>%
  rbind(tempB) %>%
  filter(!is.na(is_site_1))  %>%
  group_by(treatment, gene, is_site_1) %>%
  mutate(num_reads = length(unique(read)))


plot2 <- ggplot(tempB,
       aes(x=position56609455, color=treatment)) +
  stat_ecdf() +
  ylab("Fraction of reads") +
  xlab("Prob psU at\ndownstream site") +
  labs(color="prob psU at\nupstream site") +
  theme(axis.title.y = element_blank(),
        axis.text.y = element_blank()) +
  ggtitle("MT2A") +
  facet_grid(rows=vars(is_site_1))

figure_legend <- get_legend(plot2+theme(legend.position = "bottom",
                                     legend.background = element_blank()) )

figure <- grid.arrange(plot1, plot2 + theme(legend.position = "none"), figure_legend,  layout_matrix = layout)

tempA1 <- filter(tempA, is_site_1=="high")
tempB1 <- filter(tempB, is_site_1=="high")

pairwise_ks_test(tempA1$position134765835, tempA1$treatment, warning = -1, alternative="two.sided")
pairwise_ks_test(tempB1$position56609455,  tempB1$treatment, warning = -1, alternative="two.sided")


```

###Attempt correlation matrix
```{r}

asdf <- filter(temp3, mu_difference >= 0.4, 
               lambda_1 >= 0.01 & lambda_2 >= 0.01,
               coverage >= 0.5
               )

temp4 <- temp2 %>%
  filter(#gene == GENE,
         # treatment == "untreated"
         ) %>%
  filter(position %in% asdf$position) %>%
  mutate(readX = paste0(treatment, read)) 

attempt_cor = temp4 %>%
  ungroup() %>%
  filter(gene =="WARS1",
         treatment=="gamma") %>%
  select(position, prob_psU, read) %>%
  pivot_wider(names_from = c("position"), values_from = c(prob_psU)) %>%
  select(-read)
  
res <- cor(attempt_cor, use="pairwise.complete.obs", method="spearman") %>%
  melt() %>%
  mutate(Var1 = paste0("position",Var1),
         Var2 = paste0("position",Var2))


ggplot(res,
       aes(x=Var1, y=Var2, fill=value)) +
  geom_tile() +
  scale_fill_gradient2(low="blue", mid="white",midpoint=0, high="red") +
  theme(axis.text.x = element_text(angle=90))

# heatmap(x=res, symm = TRUE)

```


#_
#BIG LEAP
##Compute populations.
```{r}
# GENE = "B2M"
PILEUP = 1000

temp1 <- TGS_single_read %>%
  group_by(gene, position, treatment) %>%
  mutate(count=1) %>%
  summarise(mean_psU = mean(prob_psU, na.rm=T) ,
            var_psU = var(prob_psU, na.rm=T),
            pileup = sum(count) ) %>%
  filter(pileup >= PILEUP,
         # gene == GENE
         ) %>%
  mutate(positionX = paste0(treatment, position))


temp2 <- TGS_single_read %>%
  mutate(positionX = paste0(treatment, position)) %>%
  group_by(gene, treatment) %>%
  mutate(sample_pileup = length(unique(read))) %>%
  filter(positionX %in% c(temp1$positionX)) %>%
  select(-positionX)

#A function for mixture modeling
misture_model <- function(df){
  x1 = length(df$prob_psU)
  x2 = median(df$sample_pileup)
  # print(x1)
  asdf <- normalmixEM(df$prob_psU, fast=T,
                      maxit = 50,
                      k=2, arbmean=T,
                      # lambda=c(0.9, 0.1),
                      # mu = c(0,1),
                      # arbvar=T,
                      verb=F
                      )
  
  output = data.frame(mu_difference = abs(c(asdf$mu[2] - asdf$mu[1])),
                      lambda_1 = asdf$lambda[1], 
                      lambda_2 = asdf$lambda[2],
                      coverage = x1 / x2
                      )
  return(output)
}


set.seed(123)
temp3 <- temp2 %>%
  group_by(treatment, position, gene) %>%
  do(misture_model(.))

```

##Filter for good sites
```{r}


asdf <- filter(temp3, mu_difference >= 0.5, 
               lambda_1 >= 0.1 & lambda_2 >= 0.1,
               coverage >= 0.75
               )


temp4 <- temp2 %>%
  filter(#gene == GENE,
         # treatment == "untreated"
         ) %>%
  filter(position %in% asdf$position) %>%
  mutate(readX = paste0(treatment, read)) 

temp4 <- temp4 %>%
  select(-file_name) #%>%
  # mutate(position = paste0("position", position)) 


# temp4$position <- factor(temp4$position)

```

##For limited sites, use populations to filter reads for 95% confidence psU or U
```{r}
discretize_psU <- function(df){
  #Fit the distributions
  asdf <- normalmixEM(df$prob_psU, fast=T,
                      maxit = 50,
                      k=2,
                      arbmean=T,
                      # lambda=c(0.9, 0.1),
                      # mu = c(0.1,0.9),
                      # arbvar=T,
                      verb=F
                      )
  
  output <- data.frame(asdf$posterior)
  output$read = df$read
  
  output <- output %>%
    full_join(df, ., by=c("read")) %>%
    mutate(mu1 = asdf$mu[1], 
           mu2 = asdf$mu[2])
  
  return(output)
}


zxcv <- temp4 %>%
  # filter(gene=="B2M", treatment=="beta") %>%
  group_by(treatment, gene, position) %>%
  do(discretize_psU(.)) %>%
  mutate(site_call = ifelse(comp.1 >= 0.95 & mu1 <= mu2, "U",
                            ifelse(comp.1 >= 0.95 & mu1 > mu2, "psU",
                                   ifelse(comp.2 >= 0.95 & mu2 <= mu1, "U",
                                          ifelse(comp.2 >= 0.95 & mu2 > mu1, "psU",
                                                 "ambiguious")))))

# ggplot(zxcv, aes(x=prob_psU, color = site_call)) +
#   stat_ecdf() +
#   facet_wrap(~position)


```

##Comute KS for all other sites between psU and U populations.
```{r}

iterate_position <- function(df) {
  working <- unnest(df)
  working <- working %>%
    group_by(position) %>%
    nest() %>%
    # filter(position ==134765877) %>% #Esential for debugging
    do(comput_KS(., working)) %>%
    mutate(gene = df$gene,
           treatment = df$treatment)
  # print(head(working))
  return(working)
}

comput_KS <- function(df, data) {
  U_site <- filter(unnest(df), site_call=="U")
  psU_site <- filter(unnest(df), site_call=="psU")
  

  working <- data %>%
    mutate(parent_position = df$position,
           parent_site = ifelse(read %in% U_site$read, "U",
                                ifelse(read %in% psU_site$read, "psU", NA))) %>%
    ungroup() %>%
    filter(!is.na(parent_site)) %>%
    select(position, parent_site, prob_psU, parent_position)
  
  #Abort if no data here
  if( (length(filter(working, working$parent_site =="U")$position) < 100) |
      (length(filter(working, working$parent_site =="psU")$position) < 100)
  ) {
    print(df$position)
    aborted <- data.frame(parent_position = c(df$position),
           position =c(NA),
           p_value = c(NA)) %>%
      group_by(parent_position) %>%
      nest()
    return(aborted)
  
  }
  
  # print(head(working))
  working <- working %>%
    pivot_wider(names_from = c("parent_site"), values_from = c("prob_psU")) %>%
    group_by(parent_position, position) %>%
    summarise(p_value = ks.test(unlist(U), unlist(psU),
                                alternative = "two.sided",
                                exact =T)$p.value,
              statistic = ks.test(unlist(U), unlist(psU),
                                alternative = "two.sided",
                                exact =T)$statistic,
              
              ) %>%
    nest()
  return(working)
}
  
# set.seed(123)
zxcvb <- zxcv %>%
  filter(#position == 134765877,
         # gene=="IFITM3"
         # treatment=="beta"
         ) %>%
  filter(site_call %in% c("U","psU")) %>%
  select(-sample_pileup, -comp.1, -comp.2, -mu1, -mu2, -readX) %>%
  group_by(treatment, gene) %>%
  nest() %>%
  do(iterate_position(.)) %>% #%>%#iterate over each sample; position iterations inside
  select(-position) %>%
  unnest() 
  
zxcvb <- zxcvb %>%
  filter(!is.na(position))


#P-value figure
PLOTTING <- zxcvb %>%
  filter(gene=="B2M")
PLOTTING$treatment <- factor(PLOTTING$treatment, levels = c("untreated","beta","gamma"))
ggplot(PLOTTING,
       aes(x=factor(parent_position), y=factor(position), fill=log(p_value))) +
  geom_tile() +
  scale_fill_gradient2(high="grey60", mid="grey90", low="red", midpoint=-2, limits=c(-10,0), 
                       na.value = "blue") +
  geom_tile(data=filter(PLOTTING, log(p_value) <= -10), fill="darkred") +
  theme(axis.text.x = element_text(angle=90, vjust=0.5)) +
  facet_grid(rows=vars(gene),
             cols=vars(treatment))


#color by test statistic value
PLOTTING <- zxcvb %>%
  filter(gene =="B2M") %>%
  filter(p_value <= 0.01,
         # statistic <= -0.01
         )
PLOTTING$treatment <- factor(PLOTTING$treatment, levels = c("untreated","beta","gamma"))
ggplot(PLOTTING,
       aes(x=factor(parent_position), y=factor(position), fill=(statistic))) +
  geom_tile() +
  scale_fill_gradient2(high="blue", mid="grey90", low="red", midpoint=0,
                       # limits=c(-10,10),
                       na.value = "blue") +
  # geom_tile(data=filter(PLOTTING, log(p_value) <= -10), fill="darkred") +
  theme(axis.text.x = element_text(angle=90, vjust=0.5)) +
  facet_grid(rows=vars(gene),
             cols=vars(treatment))

```


#_
#Paper figure
##Single read
###B2M
```{r}
GENE = "B2M"
PILEUP = 1000

SITE1 = 44717894
SITE2 = 44717898
SITE3 = 44718109
SITE4 = 44718053

temp1 <- TGS_single_read %>%
  group_by(gene, position, treatment) %>%
  mutate(count=1) %>%
  summarise(mean_psU = mean(prob_psU, na.rm=T) ,
            var_psU = var(prob_psU, na.rm=T),
            pileup = sum(count) ) %>%
  filter(pileup >= PILEUP,
         # gene == GENE
         ) %>%
  mutate(positionX = paste0(treatment, position))

temp2 <- TGS_single_read %>%
  mutate(positionX = paste0(treatment, position)) %>%
  group_by(gene, treatment) %>%
  mutate(sample_pileup = length(unique(read))) %>%
  filter(positionX %in% c(temp1$positionX)) %>%
  select(-positionX)



tempA <- temp2 %>%
  filter(gene=="B2M",
         treatment=="untreated") %>%
  filter(position %in% c(44717894, 44717898)) %>%
  mutate(position = paste0("position", position)) %>%
  pivot_wider(names_from = c("position"), values_from = c("prob_psU")) %>%
  mutate(is_site_1 = "all")

tempA <- tempA %>%
  mutate(is_site_1 = ifelse(position44717894 >= 0.5, "high", "low")) %>%
  rbind(tempA) %>%
  filter(!is.na(is_site_1)) 

plot1 <- ggplot(tempA,
       aes(x=position44717898, color=is_site_1)) +
  stat_ecdf(size=1) +
  scale_color_manual(values=c("all"="grey", "high"="blue", "low"="lightblue")) +
  ylab("Fraction of reads") +
  xlab("Prob psU at\ndownstream site") +
  labs(color="prob psU at\nupstream site") +
  theme(legend.position = "None") +
  ggtitle(paste0(GENE," ",SITE1, " ", SITE2))


tempB <- temp2 %>%
  filter(gene=="B2M",
         treatment=="untreated") %>%
  filter(position %in% c(44718109, 44718053)) %>%
  mutate(position = paste0("position", position)) %>%
  pivot_wider(names_from = c("position"), values_from = c("prob_psU")) %>%
  mutate(is_site_1 = "all")

tempB <- tempB %>%
  mutate(is_site_1 = ifelse(position44718109 >= 0.5, "high", "low")) %>%
  rbind(tempB) %>%
  filter(!is.na(is_site_1))  %>%
  group_by(treatment, gene, is_site_1) %>%
  mutate(num_reads = length(unique(read)))


plot2 <- ggplot(tempB,
       aes(x=position44718053, color=is_site_1)) +
  stat_ecdf(size=1) +
  scale_color_manual(values=c("all"="grey", "high"="blue", "low"="lightblue")) +
  ylab("Fraction of reads") +
  xlab("Prob psU at\ndownstream site") +
  labs(color="prob psU at\nupstream site") +
  theme(axis.title.y = element_blank(),
        axis.text.y = element_blank()) +
  ggtitle(paste0(GENE," ",SITE3, " ", SITE4))

figure_legend <- get_legend(plot2+theme(legend.position = "bottom",
                                     legend.background = element_blank()) )
layout <- rbind(c(1,2), 
                c(1,2),
                c(1,2),
                c(3,3))
figure <- grid.arrange(plot1, plot2 + theme(legend.position = "none"), figure_legend,  layout_matrix = layout)

path="./"
figure_name="single_read_2_site_distributions"
width=4
height=4
scale=1.5
dpi=600
# ggsave(paste0(path, figure_name,".svg"), 
#        plot = figure,
#        scale = scale,
#        dpi = dpi,
#        width = width, 
#        height = height)

ggsave(paste0(path, figure_name,".png"),
       plot = figure,
       scale = scale,
       dpi = dpi,
       width = width,
       height = height)

```

###B2M with posterior prob
```{r}
GENE = "B2M"
PILEUP = 1000

SITE1 = 44717894
SITE2 = 44717898
SITE3 = 44718109
SITE4 = 44718053

# GENE = "B2M"
PILEUP = 1000

temp1 <- TGS_single_read %>%
  group_by(gene, position, treatment) %>%
  mutate(count=1) %>%
  summarise(mean_psU = mean(prob_psU, na.rm=T) ,
            var_psU = var(prob_psU, na.rm=T),
            pileup = sum(count) ) %>%
  filter(pileup >= PILEUP,
         gene == GENE
         ) %>%
  mutate(positionX = paste0(treatment, position))


temp2 <- TGS_single_read %>%
  mutate(positionX = paste0(treatment, position)) %>%
  group_by(gene, treatment) %>%
  mutate(sample_pileup = length(unique(read))) %>%
  filter(positionX %in% c(temp1$positionX)) %>%
  select(-positionX)

#A function for mixture modeling
misture_model <- function(df){
  x1 = length(df$prob_psU)
  x2 = median(df$sample_pileup)
  # print(x1)
  asdf <- normalmixEM(df$prob_psU, fast=T,
                      maxit = 50,
                      k=2, arbmean=T,
                      # lambda=c(0.9, 0.1),
                      # mu = c(0,1),
                      # arbvar=T,
                      verb=F
                      )
  
  output = data.frame(mu_difference = abs(c(asdf$mu[2] - asdf$mu[1])),
                      lambda_1 = asdf$lambda[1], 
                      lambda_2 = asdf$lambda[2],
                      coverage = x1 / x2
                      )
  return(output)
}


set.seed(123)
temp3 <- temp2 %>%
  group_by(treatment, position, gene) %>%
  do(misture_model(.))



asdf <- filter(temp3, mu_difference >= 0.5, 
               lambda_1 >= 0.1 & lambda_2 >= 0.1,
               coverage >= 0.75
               )
temp4 <- temp2 %>%
  filter(#gene == GENE,
         # treatment == "untreated"
         ) %>%
  filter(position %in% asdf$position) %>%
  mutate(readX = paste0(treatment, read)) 
temp4 <- temp4 %>%
  select(-file_name) 



discretize_psU <- function(df){
  #Fit the distributions
  asdf <- normalmixEM(df$prob_psU, fast=T,
                      maxit = 50,
                      k=2,
                      arbmean=T,
                      # lambda=c(0.9, 0.1),
                      # mu = c(0.1,0.9),
                      # arbvar=T,
                      verb=F
                      )
  
  output <- data.frame(asdf$posterior)
  output$read = df$read
  
  output <- output %>%
    full_join(df, ., by=c("read")) %>%
    mutate(mu1 = asdf$mu[1], 
           mu2 = asdf$mu[2])
  
  return(output)
}


zxcv <- temp4 %>%
  # filter(gene=="B2M", treatment=="beta") %>%
  group_by(treatment, gene, position) %>%
  do(discretize_psU(.)) %>%
  mutate(site_call = ifelse(comp.1 >= 0.95 & mu1 <= mu2, "U",
                            ifelse(comp.1 >= 0.95 & mu1 > mu2, "psU",
                                   ifelse(comp.2 >= 0.95 & mu2 <= mu1, "U",
                                          ifelse(comp.2 >= 0.95 & mu2 > mu1, "psU",
                                                 "ambiguious")))))

# ggplot(zxcv, aes(x=prob_psU, color = site_call)) +
#   stat_ecdf() +
#   facet_wrap(~position)



tempA <- zxcv %>%
  filter(gene=="B2M",
         treatment=="untreated")
tempA <- tempA %>%
  ungroup() %>%
  filter(position == SITE1) %>%
  mutate(parent_site_call = site_call) %>%
  select(gene, read, treatment, parent_site_call) %>%
  full_join(., filter(tempA, position ==SITE2), by=c("gene", "read", "treatment")) %>%
  filter(!is.na(parent_site_call),
         parent_site_call != "ambiguious") %>%
  mutate(position = paste0("position", position)) 

plot1 <- ggplot(tempA,
       aes(x=prob_psU, color=parent_site_call)) +
  stat_ecdf(size=1) +
  scale_color_manual(values=c("psU"="blue", "U"="lightblue")) +
  ylab("Fraction of reads") +
  xlab("Prob psU at\ndownstream site") +
  labs(color="prob psU at\nupstream site") +
  theme(legend.position = "None") +
  ggtitle(paste0(GENE," ",SITE1, " ", SITE2)) +
  theme(plot.title = element_text(size = 12))


tempB <- zxcv %>%
  filter(gene=="B2M",
         treatment=="untreated")
tempB <- tempB %>%
  ungroup() %>%
  filter(position == SITE3) %>%
  mutate(parent_site_call = site_call) %>%
  select(gene, read, treatment, parent_site_call) %>%
  full_join(., filter(tempB, position ==SITE4), by=c("gene", "read", "treatment")) %>%
  filter(!is.na(parent_site_call),
         parent_site_call != "ambiguious") %>%
  mutate(position = paste0("position", position)) 


plot2 <- ggplot(tempB,
       aes(x=prob_psU, color=parent_site_call)) +
  stat_ecdf(size=1) +
  scale_color_manual(values=c("psU"="blue", "U"="lightblue")) +
  ylab("Fraction of reads") +
  xlab("Prob psU at\ndownstream site") +
  labs(color="prob psU at\nupstream site") +
  # theme(legend.position = "None") +
  # theme(axis.title.y = element_blank(),
        # axis.text.y = element_blank()) +
  ggtitle(paste0(GENE," ",SITE3, " ", SITE4)) +
  theme(plot.title = element_text(size = 12))

figure_legend <- get_legend(plot2+theme(legend.position = "bottom",
                                     legend.background = element_blank()) )
layout <- rbind(c(1,2), 
                c(1,2),
                c(1,2),
                c(3,3))
figure <- grid.arrange(plot1, plot2 + theme(legend.position = "none"), figure_legend,  layout_matrix = layout)

path="./"
figure_name="single_read_2_site_distributions"
width=4
height=4
scale=1.5
dpi=600
# ggsave(paste0(path, figure_name,".svg"), 
#        plot = figure,
#        scale = scale,
#        dpi = dpi,
#        width = width, 
#        height = height)

ggsave(paste0(path, figure_name,".png"),
       plot = figure,
       scale = scale,
       dpi = dpi,
       width = width,
       height = height)

```


#_
Also interesting: sites that gain or lose psU with treatment

#_
#_


